homeassistant/components/watts/coordinator.py

```diff
@@ -0,0 +1,247 @@
+
+from .const import (
+    DISCOVERY_INTERVAL_MINUTES,
+    DOMAIN,
+    FAST_POLLING_INTERVAL_SECONDS,
+    UPDATE_INTERVAL_SECONDS,
+)
+
+if TYPE_CHECKING:
+    from . import WattsVisionRuntimeData
+
+    type WattsVisionConfigEntry = ConfigEntry[WattsVisionRuntimeData]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class WattsVisionThermostatData:
+    """Data class for thermostat device coordinator."""
+
+    thermostat: ThermostatDevice
+
+
+class WattsVisionHubCoordinator(DataUpdateCoordinator[dict[str, Device]]):
+    """Hub coordinator for bulk device discovery and updates."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client: WattsVisionClient,
+        config_entry: WattsVisionConfigEntry,
+    ) -> None:
+        """Initialize the hub coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=UPDATE_INTERVAL_SECONDS),
+            config_entry=config_entry,
+        )
+        self.client = client
+        self._last_discovery: datetime | None = None
+        self.previous_devices: set[str] = set()
+
+    async def async_setup(self) -> None:
+        """Set up the coordinator by discovering devices."""
+        try:
+            devices_list = await self.client.discover_devices()
+        except WattsVisionAuthError as err:
+            raise ConfigEntryAuthFailed("Authentication failed") from err
+        except (
+            WattsVisionConnectionError,
+            WattsVisionTimeoutError,
+            WattsVisionDeviceError,
+            WattsVisionError,
+            ConnectionError,
+            TimeoutError,
+            ValueError,
+        ) as err:
+            raise ConfigEntryNotReady("Failed to discover devices") from err
+
+        devices = {device.device_id: device for device in devices_list}
+        _LOGGER.info("Initial discovery completed with %d devices", len(devices))
+        self.async_set_updated_data(devices)
+
+        self._last_discovery = datetime.now()
+
+    async def _async_update_data(self) -> dict[str, Device]:
+        """Fetch data and periodic device discovery."""
+        now = datetime.now()
+
+        if self._last_discovery is None or now - self._last_discovery >= timedelta(
+            minutes=DISCOVERY_INTERVAL_MINUTES
+        ):
+            try:
+                devices_list = await self.client.discover_devices()
+            except WattsVisionAuthError as err:
+                _LOGGER.warning("Periodic discovery failed with auth error: %s", err)
+                raise ConfigEntryAuthFailed("Authentication failed") from err
+            except (
+                WattsVisionConnectionError,
+                WattsVisionTimeoutError,
+                WattsVisionDeviceError,
+                WattsVisionError,
+                ConnectionError,
+                TimeoutError,
+                ValueError,
+            ) as err:
```

Comment:
Why do we have simialr code for discover_devices in async_setup en _async_update_data?